pkg/sql/tenant.go

```diff
@@ -211,38 +234,81 @@ func ActivateTenant(ctx context.Context, execCfg *ExecutorConfig, txn *kv.Txn, t
+		RequestHeader: roachpb.RequestHeader{Key: prefix, EndKey: prefixEnd},
+	})
+
+	return errors.Wrapf(execCfg.DB.Run(ctx, b), "clearing tenant %d data", tenID)
+}
+
 // DestroyTenant implements the tree.TenantOperator interface.
-func DestroyTenant(ctx context.Context, execCfg *ExecutorConfig, txn *kv.Txn, tenID uint64) error {
+func (p *planner) DestroyTenant(ctx context.Context, tenID uint64) error {
 	const op = "destroy"
-	if err := rejectIfCantCoordinateMultiTenancy(execCfg.Codec, op); err != nil {
+	if err := rejectIfCantCoordinateMultiTenancy(p.execCfg.Codec, op); err != nil {
 		return err
 	}
 	if err := rejectIfSystemTenant(tenID, op); err != nil {
 		return err
 	}
 
 	// Retrieve the tenant's info.
-	info, err := getTenantRecord(ctx, execCfg, txn, tenID)
+	info, err := getTenantRecord(ctx, p.execCfg, p.txn, tenID)
 	if err != nil {
 		return errors.Wrap(err, "destroying tenant")
 	}
 
+	if info.State == descpb.TenantInfo_DROP {
+		return errors.Errorf("tenant %d is already in state DROP", tenID)
+	}
+
 	// Mark the tenant as dropping.
 	info.State = descpb.TenantInfo_DROP
-	if err := updateTenantRecord(ctx, execCfg, txn, info); err != nil {
-		return errors.Wrap(err, "destroying tenant")
+	return errors.Wrap(updateTenantRecord(ctx, p.execCfg, p.txn, info), "destroying tenant")
+}
+
+// GCTenant implements the tree.TenantOperator interface.
+func (p *planner) GCTenant(ctx context.Context, tenID uint64) error {
+	if !p.ExtendedEvalContext().TxnImplicit {
+		return errors.Errorf("gc_tenant cannot be used inside a transaction")
 	}
 
-	// TODO(nvanbenschoten): actually clear tenant keyspace. We don't want to do
-	// this synchronously in the same transaction, because we could be deleting
-	// a very large amount of data. Instead, we should kick off a job that picks
-	// up the DROP state of the tenant, clears the tenant keyspace, and deletes
-	// the row in the system.tenants table. Tracked in #48775.
+	const op = "gc"
+	if err := rejectIfCantCoordinateMultiTenancy(p.execCfg.Codec, op); err != nil {
+		return err
+	}
+	if err := rejectIfSystemTenant(tenID, op); err != nil {
+		return err
+	}
 
-	return nil
-}
+	if err := p.execCfg.DB.Txn(ctx, func(ctx context.Context, txn *kv.Txn) error {
+		info, err := getTenantRecord(ctx, p.execCfg, p.txn, tenID)
+		if err != nil {
+			return errors.Wrap(err, "get tenant record")
+		}
+		if info.State != descpb.TenantInfo_DROP {
+			return errors.Errorf("tenant %d is not in state DROP", tenID)
+		}
+		return nil
+	}); err != nil {
+		return err
+	}
 
-// DestroyTenant implements the tree.TenantOperator interface.
-func (p *planner) DestroyTenant(ctx context.Context, tenID uint64) error {
-	return DestroyTenant(ctx, p.ExecCfg(), p.Txn(), tenID)
+	if err := ClearTenant(ctx, p.ExecCfg(), tenID); err != nil {
+		return errors.Wrap(err, "clear tenant")
+	}
+
+	return errors.Wrapf(
+		DeleteTenantRecord(ctx, p.ExecCfg(), p.txn, tenID), "deleting tenant %d record", tenID,
+	)
```

Comment:
should we `return p.txn.Commit()` before we let go of txn? we did non-txn'l side effects (clearing range) so I'm wary of leaving p.txn uncommitted, but I might just be paranoid. 

Alternatively, we could just fire up our own txn again and completely ignore p.Txn (other than asserting it is implicit) which might be cleaner if we do the method/function re-org I mention above, because then this function isn't passed a txn at all which makes clear that it isn't transactional.